### PR TITLE
Remove legacy login handling

### PR DESCRIPTION
**Pull request by DanVanAtta**
## Overview
Remove code in login that handled previous clients. Latest lobby will only be available to latest client. Sending login requests from latest client, this update removes server side paths that are now impossible. With those code paths removed, failing tests were removed.

## Functional Changes
None, legacy clients will not be able to connect to latest lobby; 

## Manual Testing Performed
- Ran lobby code through a debugger while logging in with a latest client and was able to login.

### CODE REVIEW

File: lobby/src/main/java/org/triplea/lobby/server/db/UserController.java [hunk 1]
```diff
@@ -97,17 +97,14 @@ private static String getPasswordColumn(final HashedPassword hashedPassword) {
   @Override
   public void createUser(
       final String name, final String email, final HashedPassword hashedPassword) {
-    Preconditions.checkState(hashedPassword.isHashedWithSalt());
```
Comment (DanVanAtta) [inline lobby/src/main/java/org/triplea/lobby/server/db/UserController.java]: This condition is already checked in caller code.

File: lobby/src/main/java/org/triplea/lobby/server/db/UserController.java [hunk 2]
```diff
@@ -97,17 +97,14 @@ private static String getPasswordColumn(final HashedPassword hashedPassword) {
...
-
     try (Connection con = connection.get();
         PreparedStatement ps =
             con.prepareStatement(
-                "insert into lobby_user (username, password, bcrypt_password, email) "
```
Comment (DanVanAtta) [inline lobby/src/main/java/org/triplea/lobby/server/db/UserController.java]: Latest code, when creating a user, should always use a bcrypted password.

File: lobby/src/main/java/org/triplea/lobby/server/db/UserController.java [hunk 3]
```diff
@@ -122,27 +119,13 @@ public void createUser(
   @Override
   public boolean login(final String username, final HashedPassword hashedPassword) {
     try (Connection con = connection.get()) {
-      if (hashedPassword.isHashedWithSalt()) {
```
Comment (DanVanAtta) [inline lobby/src/main/java/org/triplea/lobby/server/db/UserController.java]: This if branch is never executed with latest client, the 'else' is preserved

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java [hunk 4]
```diff
@@ -186,59 +186,26 @@ private static String getBanDurationBreakdown(final Timestamp stamp) {
...
     if (hashedPassword == null) {
-      return errorMessage;
+      return ErrorMessages.AUTHENTICATION_FAILED;
     }
+
     if (RsaAuthenticator.canProcessResponse(response)) {
       return rsaAuthenticator.decryptPasswordForAction(
           response,
           pass -> {
-            final String legacyHashedPassword =
```
Comment (DanVanAtta) [inline lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: latest client does not send key 'HASHED_PASSWORD', legacyPassword is always null.

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java [hunk 5]
```diff
@@ -186,59 +186,26 @@ private static String getBanDurationBreakdown(final Timestamp stamp) {
...
-                response.get(LobbyLoginResponseKeys.HASHED_PASSWORD);
             if (hashedPassword.isBcrypted()) {
-              if (database.getUserDao().login(username, new HashedPassword(pass))) {
```
Comment (DanVanAtta) [inline lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: The outer if statement is always executed, the inner if statement is always false as `legacyHashedPassword` is always null. The if statement on 201 is preserved.

File: lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java [hunk 6]
```diff
@@ -61,17 +61,16 @@ void testCreateDupe() {
...
     controller.updateUser(
         user, generateEmailAddress(user), new HashedPassword(bcrypt(obfuscate(password))));
-    assertTrue(controller.login(user, new HashedPassword(password)));
     assertTrue(controller.login(user, new HashedPassword(obfuscate(password))));
   }
 
   @Test
   void testUpdate() throws Exception {
-    final String user = newUserWithMd5CryptHash();
+    final String user = newUserWithBCryptHash();
```
Comment (RoiEXLab) [inline lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java]: I believe this test was in place to check if the legacy passwords were migrated correctly, so it can be simplified

File: lobby/src/test/java/org/triplea/lobby/server/login/LobbyLoginValidatorIntegrationTest.java [hunk 7]
```diff
@@ -136,7 +118,9 @@ void testAnonymousLogin() {
     // create a user, verify we can't login with a username that already exists
     // we should not be able to login now
     assertNotNull(
-        generateChallenge(new HashedPassword(md5Crypt("foo"))).apply(challenge -> response));
+        generateChallenge(
+                new HashedPassword(BCrypt.hashpw(hashPasswordWithSalt("foo"), BCrypt.gensalt())))
+            .apply(challenge -> response));
```
Comment (RoiEXLab) [inline lobby/src/test/java/org/triplea/lobby/server/login/LobbyLoginValidatorIntegrationTest.java]: I could've sworn that there was a similar test case would exist somewhere in this file already (one for legacy, one for "new"), but it doesn't seem like that.

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java [hunk 8]
```diff
@@ -186,59 +186,26 @@ private static String getBanDurationBreakdown(final Timestamp stamp) {
...
-                  .updateUser(
-                      username,
-                      email,
-                      new HashedPassword(BCrypt.hashpw(pass, bcryptSaltGenerator.get())));
-              return null;
+              return database.getUserDao().login(username, new HashedPassword(pass))
+                  ? null
+                  : ErrorMessages.AUTHENTICATION_FAILED;
             } else {
-              return errorMessage;
```
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: To clarify:
Are you planning to drop the other password column at some point or keep it for now?

[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: Only latest clients will be able to connect to the next lobby. The code removed should then be simply dead code.

With luck we've got it so that any accounts with only md5 encrypted passwords will still be able to login, but looking at the code removed, it doesn't quite look like we'll do the upgrade of password for the case of latest client still using md5.

Push comes to shove we can ask users to use the 'forgot password' feature coming along to reset their password. Bottom line though is the code removed should be dead code.

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java [hunk 9]
```diff
@@ -186,59 +171,26 @@ private static String getBanDurationBreakdown(final Timestamp stamp) {
...
+              return "Badly hashed password in client request";
             }
           });
+    } else {
+      return "Badly formatted client request";
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: Generally assuming that it's only latest clients; would be controlled by the lobby server properties file and we'll increment the 'min' triplea version so old clients will get notified they should upgrade. So, assuming the latter case, that a latest triplea is being used (which is the crux of this PR, we'll send latest clients to latest lobby, which means any code to deal with legacy clients is dead code).

It's been some time in the works that the next release will be a "parallel" lobby, we'll have a latest instance set up and keep the old one running while we migrate users. We locked in that decision with several non-compatible database updates.

File: lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java [hunk 10]
```diff
@@ -27,7 +27,7 @@
   void testCreate() throws Exception {
     final int startCount = getUserCount();
 
-    newUserWithMd5CryptHash();
+    newUserWithBCryptHash();
     assertEquals(getUserCount(), startCount + 1);
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java]: Hmm, sorry to not spot it, but these tests look like they're doing slightly different things:
```
  @Test
  void testCreate() throws Exception {
    final int startCount = getUserCount();

    newUserWithBCryptHash();
    assertEquals(getUserCount(), startCount + 1);

    newUserWithBCryptHash();
    assertEquals(getUserCount(), startCount + 2);
  }

  @Test
  void testGet() {
    final String user = newUserWithBCryptHash();
    assertEquals(generateEmailAddress(user), controller.getUserEmailByName(user));
  }
```

File: lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java [hunk 11]
```diff
@@ -36,19 +36,19 @@ void testCreate() throws Exception {
...
   void testGet() {
-    final String user = newUserWithMd5CryptHash();
+    final String user = newUserWithBCryptHash();
     assertEquals(generateEmailAddress(user), controller.getUserEmailByName(user));
   }
 
   @Test
   void testDoesUserExist() {
-    assertTrue(controller.doesUserExist(newUserWithMd5CryptHash()));
+    assertTrue(controller.doesUserExist(newUserWithBCryptHash()));
```
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java]: Updated: 7bd4a42d6

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: I mean one option would just be to grab all the emails we have of people that don't have an updated password, send them a mail (manually over some mass email service or something) they have to login using the latest stable, otherwise their account will be deleted on October 1st (or other planned upgrade date).
Then we could safely drop the column, rename the current column potentially, and remove all users that don't have an updated password column. They have been warned.

File: lobby/src/test/java/org/triplea/lobby/server/login/LobbyLoginValidatorIntegrationTest.java (hunk 7 above)
Comment (DanVanAtta) [reply lobby/src/test/java/org/triplea/lobby/server/login/LobbyLoginValidatorIntegrationTest.java]: Thanks for checking, having slightly redundant tests for now I think is okay:
- there are not too many redundant tests if any
- the extra coverage is not a bad trade-off for some potential redundancies, particularly for a section of code that is 'risky' (difficult to get absolutely correct and very important that it is correct)

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: > To clarify:
Are you planning to drop the other password column at some point or keep it for now?

There is a select with coelesce on the old password column; it is not clear to me that we can drop it.

File: lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java (hunk 6 above)
Comment (DanVanAtta) [reply lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java]: That original intent is not clear. I'm afraid you'll need to follow up on this if it is not simplified in other updates first.

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: True, but if we'd just remove all users without a migrated password, the coelesce wouldn't be needed

[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: I'm hesitant to remove the column and coelesce. This PR was me going through with a debugger and only removing dead code branches.

For users with an email, they'll be able to use 'forgot password' and we won't need to drop their user row. Otherwise users that have not upgraded will have their game ***freeze on startup***. We've a latent bug in the 3xxx version where if the 'triplea min' version is incremented their game freezes. There is a lot of motivation to get users on 7xxx versions of the game.

Ultimately I suspect users without an email or a bcrypted password would be dropped. For now delaying that decision is good, it's not strictly needed to make that call just yet, but it seems to be coming soon.

**Comment by DanVanAtta** [comment]
@RoiEXLab please let me know if there is anything that should still be addressed. Thank you for double checking the update and thinking through its consequences with me.

File: lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java]: @DanVanAtta Sorry I should've been more clear:
I was refferring to this part of the code:
```
    newUserWithBCryptHash();
    assertEquals(getUserCount(), startCount + 1);

    newUserWithBCryptHash();
    assertEquals(getUserCount(), startCount + 2);
```
Previously the code tested if Md5Crypt and BCrypt incremented the user count individually. Now we test if BCrypt increments the counter twice.

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: Ah ok now I get what you are trying to achieve.
I can tell you with high confidence that logging in with a md5-only password won't work. The Md5Crypt challenge code you removed is crucial for this: From what I found out quite some time ago the legacy login system doesn't use Md5Crypt to actually hash passwords before putting them in the database, instead it is used to act as pseudo encryption before actually sending the password over the network.
So in order for the client to authenticate with md5 passwords, the server needs to send the salt of the password stored in the database to the client, which then Md5Crypt-hashes the password using this exact salt and sends the hash over the network which is then compared with the database on the server.

That's why I was so confused initially. For me this PR looks like it removes the critical legacy authentication functions, but doesn't remove everything related to it.

Thoughts?

[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: Looking at this code below (LobbyLoginValidator:67), there is no client code that fetches the SALT key from challenge:

```
  private Map<String, String> newMd5CryptAuthenticatorChallenge(final String userName) {
    final Map<String, String> challenge = new HashMap<>();
    final HashedPassword password = database.getUserDao().getLegacyPassword(userName);
    if (password != null && Strings.emptyToNull(password.value) != null) {
      challenge.put(LobbyLoginChallengeKeys.SALT, Md5Crypt.getSalt(password.value));
    }
    return challenge;
  }

```


Looking at this code (LobbyLoginValidator:278), there is no code that 'puts' the HASHED_PASSWORD key into a challenge. All accessors to that property are 'get' and are on the server.

```
   final HashedPassword password =
        new HashedPassword(response.get(LobbyLoginResponseKeys.HASHED_PASSWORD));
    if (RsaAuthenticator.canProcessResponse(response)) {
      return rsaAuthenticator.decryptPasswordForAction(
          response,
          pass -> {
            final HashedPassword newPass =
                new HashedPassword(BCrypt.hashpw(pass, bcryptSaltGenerator.get()));
            if (password.isHashedWithSalt()) {
              database.getUserDao().createUser(username, email, password);
              database.getUserDao().updateUser(username, email, newPass);
            } else {
              database.getUserDao().createUser(username, email, newPass);
            }
            return null;
          });
    }
```

It somewhat seems that we might have a mistake where latest clients cannot do MD5Crypt password authentication to start with. Regardless if that is the case, given all current client, the code removed seem to be impossible code paths when using a current client.


> For me this PR looks like it removes the critical legacy authentication functions, but doesn't remove everything related to it.

In short, it appears all legacy authentication functions are to support 'old' clients. If we assume a latest client with current code, those authentication mechanisms would never have been triggered to start with. Hence, moves to remove those paths.

[the remark this answers is not in the document]
Comment (RoiEXLab) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: You're right. The current client can't use that code at all, I forgot that we dropped some code on the client side quite some time ago. I even found @ssoloff 's PR https://github.com/triplea-game/triplea/pull/4262

I don't even know which engine version this represents (probably 1.9).

So there's a thing I don't quite understand: If old login handling code is removed which makes it impossible logging in using a legacy client, why keep the HashedPassword wrapper class for example that is only used to handle different types of hashes?
If this is just "step 1" of the removal then I'm ok with it, but I feel like you didn't make clear what comes next exactly in terms of refactoring (the high level explanation with the parallel lobby did help though, explains why the DB structure will be kept).

File: lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java]: So this is the only objection I have left. Otherwise this looks good

[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java]: Hmm, I'm not sure if the extra increment is that bad of a check, seems like more test coverage and not necessarily redundant. Do you feel strongly the second `startCount + 2` check should be removed?

File: lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (DanVanAtta) [reply lobby/src/main/java/org/triplea/lobby/server/login/LobbyLoginValidator.java]: >  If old login handling code is removed which makes it impossible logging in using a legacy client

Assuming I'm correct that these are impossible code paths with a current client, and given we will force a latest client for the 1.10/2.0 release, then it was already impossible (dead code).

>  I'm ok with it, but I feel like you didn't make clear what comes next exactly in terms of refactoring

Great question, TBH I'm not sure. There are some big refactoring changes that could happen, but it's too much for one PR. For now this is taking a slice off of the complexity so the code is a bit easier to work with by removing dead code. With luck password reset will be easier to implement, I spent a lot of time in this code yesterday working on the encryption pieces and eventually concluded the code needed simplification first or otherwise the risk was too high and ultimately wasting time working in an overly complicated area of the code.

It does seem there could be more iterations. With luck those would be more obvious now that this work is done. Getting to that point where such things would be move obvious is perhaps the largest goal here, not necessarily reducing this code to the simplest possible. I am tempted to spend some more time to try and improve things a bit more before picking back up password reset.

File: lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (RoiEXLab) [reply lobby/src/test/java/org/triplea/lobby/server/db/UserControllerIntegrationTest.java]: I don't have any strong feelings about it, but I feel like if we come back at this test in say 3 months, we might start thinking that there might be special behaviour for the second user creation 🤷‍♂

**Comment by DanVanAtta** [comment]
Looking at the test, it looks like checking for multiple cases is helping verify the "N+1" case, it does not bother me seeing it. I'll go ahead and hit merge then, I appreciate you working with me on this review @RoiEXLab , it was good to confirm my understanding and get the second pair of eyes and us on the same page - thank you!
